streams/src/test/java/org/apache/kafka/streams/processor/internals/TransactionManagerTest.java

```diff
@@ -0,0 +1,638 @@
+        "cluster",
+        Collections.singletonList(Node.noNode()),
+        Collections.singletonList(new PartitionInfo(topic, 0, Node.noNode(), new Node[0], new Node[0])),
+        Collections.emptySet(),
+        Collections.emptySet()
+    );
+    private final TaskId taskId = new TaskId(0, 0);
+    private final ByteArraySerializer byteArraySerializer = new ByteArraySerializer();
+    private final Map<TopicPartition, OffsetAndMetadata> offsetsAndMetadata = mkMap(
+        mkEntry(new TopicPartition(topic, 0), new OffsetAndMetadata(0L, null))
+    );
+
+    private final MockProducer<byte[], byte[]> mockProducer = new MockProducer<>(
+        cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer);
+    private final TransactionManager transactionManager = new TransactionManager(logContext, mockProducer);
+
+    private final MockProducer<byte[], byte[]> eosMockProducer = new MockProducer<>(
+        cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer);
+    private final TransactionManager eosTransactionManager = new TransactionManager(logContext, eosMockProducer, "appId", taskId);
+
+    private final ProducerRecord<byte[], byte[]> record =
+        new ProducerRecord<>(topic, 0, 0L, new byte[0], new byte[0], new RecordHeaders());
+
+    @Before
+    public void before() {
+        eosTransactionManager.initTransaction();
+    }
+
+    @Test
+    public void shouldFailIfProducerIsNull() {
+        {
+            final NullPointerException thrown = assertThrows(
+                NullPointerException.class,
+                () -> new TransactionManager(logContext, null)
+            );
+
+            assertThat(thrown.getMessage(), equalTo("producer cannot be null"));
+        }
+
+        {
+            final NullPointerException thrown = assertThrows(
+                NullPointerException.class,
+                () -> new TransactionManager(logContext, null, "appId", taskId)
+            );
+
+            assertThat(thrown.getMessage(), equalTo("producer cannot be null"));
+        }
+    }
+
+    @Test
+    public void shouldThrowIfIncorrectlyInitialized() {
+        {
+            final IllegalArgumentException thrown = assertThrows(
+                IllegalArgumentException.class,
+                () -> new TransactionManager(logContext, mockProducer, null, taskId)
+            );
+            assertThat(thrown.getMessage(), equalTo("applicationId and taskId must either be both null or both be not null"));
+        }
+
+        {
+            final IllegalArgumentException thrown = assertThrows(
+                IllegalArgumentException.class,
+                () -> new TransactionManager(logContext, mockProducer, "appId", null)
+            );
+            assertThat(thrown.getMessage(), equalTo("applicationId and taskId must either be both null or both be not null"));
+        }
+    }
+
+    // non-eos tests
+
+    // positive tests
```

Comment:
qq: not sure what does this mean.

Reply:
It means functional test, ie, we that something should work (instead of negative test that test that something should not work/throw an error). Are you fine with this or do you have a suggestion on how to rephrase it?

Reply:
Yeah this is fine.